### ISSUE TITLE
Not being redirected from Google and feature requests

### ISSUE THREAD

**Issue by victor10100111001** [libredirect/libredirect#448]
Hi everyone

**Issue:**
These are my settings in LibRedirect. However when i go to [google.com](https://google.com) i don't get redirected automatically. Does anyone know what might cause this?
<img width="565" alt="Screenshot 2022-09-17 112947" src="https://user-images.githubusercontent.com/112384942/190850110-af44a1bc-f212-4a6c-81f6-51a9b73d7645.png">

**Feature requests:**
It would be nice if you were able to automatically check the latency of every instance there is (Reddit, YouTube, Imgur, etc). Also it seems like LibRedirect is only checking one instance at a time and will only check the next instance when it has gotten a response back. This makes it take quite some time for all instances to be checked. Would it make sense to check all the instances at once?

I have also noticed that some instances just doesn't work ([example](https://lr.stilic.ml/)) or i get a warning from uBlock Origin saying the domain is malicious. What do you think about implementing a system where users could report other instances?

Lastly it would be nice if there was a way to automatically disable JavaScript on every instance. There is a constantly changing list of instances and it is a hassle to manually disable JavaScript on each of them in uBlock Origin.

Please let me know your thoughts.

Best regards,
Victor

**Comment by ManeraKai** [comment]
From [Hygna](https://codeberg.org/Hygna):
1. There is not really much point in redirecting search, just set libredirect as your default search engine. Plus, people may use something like google for reverse image search, so it would be annoying to have to disable search every time.
   
   
2. Automatically checking all instsnces by all users might just ddos some smaller instances, so we do not want that, especially if the user doesn't care about latency. Plus, some people do not want uncessecery web requests going out, as seen in some tickets.
   
   
3. Perhaps we can check multiple instances at a time, the only issue is it might be more complicated to code. Also, it's not like you have to wait for the whole time, you can go do something else in the mean time. And on top of that, the latency is an avarage calculated by doing multiple requests, which might complicate things even more.
   
   
4. If an instance goes bad, you can report them to the respective offical instance list, unless you see the list as a file called `<frontend>.json` inside `src/instances`, then open a ticket.
5. I do not think ubo has an api for setting filters, perhaps just enable the option to have all websites have no javascript by default.

**Comment by victor10100111001** [comment]
Thanks for the answer

1. Maybe there should be an option where you can select which domains (eg. bing.com or google.com) redirect to whoogle og searxng. Personally i would like to be able to use duckduckgo as my main and if i can't find what i am looking for there i would like to be able to go to google.com and then automatically be redirected to a more privacy respecting alternative.

2. 3. 4. I agree (:

5. I was thinking of LibRedirect implementing its own javascript blocking feature. I don't know how hard that would be to implement and if it would be acceptable to take some code from uBlock Origin itself?

Please let me know your thoughts.

Best regards,
Victor

**Comment by ManeraKai** [comment]
> Maybe there should be an option where you can select which domains (eg. bing.com or google.com) redirect to whoogle og searxng. Personally i would like to be able to use duckduckgo as my main and if i can't find what i am looking for there i would like to be able to go to google.com and then automatically be redirected to a more privacy respecting alternative.

For this you go to search.libredirect.invalid and you'll get redirected rather than going to google.com

**Comment by ManeraKai** [comment]
> I was thinking of LibRedirect implementing its own javascript blocking feature. I don't know how hard that would be to implement and if it would be acceptable to take some code from uBlock Origin itself?

Hmm... It's not on our current goals but a really interesting idea to implement, but I don't think it will be easy at all.

**Comment by victor10100111001** [comment]
What are the current goals for this project? Is there a list somewhere?

**Comment by ManeraKai** [comment]
from Hygna:
1. There is a feature in ddg called bangs, which allow you to quickly switch search engines. If you want, you can [suggest to add a libredirect bang](https://duckduckgo.com/bang?)
5. I feel like this is just mission creep, and will just end up biting off more then we can chew. If you want to block js by default, you can just use an extension like noscript or just change your ubo settings to block js by default.

**Comment by egad13** [comment]
> Maybe there should be an option where you can select which domains (eg. bing.com or google.com) redirect to whoogle og searxng. Personally i would like to be able to use duckduckgo as my main and if i can't find what i am looking for there i would like to be able to go to google.com and then automatically be redirected to a more privacy respecting alternative.

Seconding this as an option that would be nice to have. Being able to go to search.libredirect.invalid is a nice workaround but it would also be nice to have the option to auto-redirect some common search engines. Links to google/bing/etc are uncommon but still show up in some places.

---

In the meantime though, here's a workaround for anyone searching for one.

To automatically redirect Google and Bing to whatever search you've configured LibRedirect for:

1. Install the add-on [Redirector](https://addons.mozilla.org/firefox/addon/redirector/)
2. Create a new redirect. This one will redirect links that lead directly to search results.
    1. Set "Example URL" `https://www.google.com/search?q=example`, "Redirect to" url to `https://search.libredirect.invalid/search?q=$1`, and "Pattern Type" to Regex.
    2. Set "Include Pattern" to `https?://www\.(?:google|bing)\.(?:com|ca)/search\?[\s\S]*q=([\d\w %]*)[\s\S]*`. Be sure to add TLDs other than .com and .ca if you need them; [Google has lots of TLDs](https://infogalactic.com/info/List_of_Google_domains).
    3. Give the redirect a name and save it.
3. Create another new redirect. This one will redirect any other link under the basic Google and Bing domains.
    1. Set "Example URL" to `https://www.google.com/`, "Redirect to" url to `https://search.libredirect.invalid/`, and "Pattern Type" to Regex.
    2. Set "Include Pattern" to `https?://www\.(?:google|bing)\.(?:com|ca)`. Be sure to add TLDs other than .com and .ca if you need them.
    3. Give the redirect a name and save it.
4. Ensure the search redirect is positioned above the generic redirect in the list.
5. Test it out!

If you ever need to actually view a Google/Bing page, Redirector can easily be enabled and disabled from the toolbar.

**Comment by ManeraKai** [comment]
> Seconding this as an option that would be nice to have. Being able to go to search.libredirect.invalid is a nice workaround but it would also be nice to have the option to auto-redirect some common search engines. Links to google/bing/etc are uncommon but still show up in some places.

This is not a priority rn, but I think we should link your way of doing it in Redirector in the website. @IkelAtomig?

**Comment by IkelAtomig** [comment]
@egad13 What does your solution actually do ?

Could you ELI5 ?

If I cold understand, I might add your solution to the site's FAQ.

**Comment by Vendillah** [comment]
For people that just want to quickly redirect [google.com](google.com): there is also the outdated [Privacy Redirect](https://github.com/SimonBrazell/privacy-redirect) (PR), which does redirect google.com to a user defined "instance", such as ddg. You can disable all the other redirects in its settings (which is what I'm doing). I do agree with the statement that it can be annoying that [images.google.com](images.google.com) gets redirected; for those very rare cases that I need it, I disable it briefly.

What's more considerable in my case is that in some situations Google redirects you a couple of times on certain actions (e.g. before or after a google-account login _FOR SOME REASON?!_) before landing on a page, and the url "google.com" happens to be captured by PR. I think one should also think about what to do in those situations on proposing/implementing a search redirect like PR does.

Also, I would like to suggest for this extension to inform the user that the options under "search" (engine) don't work out of the box (but instead only when LibRedirect is used as the search engine?), because it's not clear yet, and I myself thought it was simply broken for the moment and had been waiting for a fixing update to finally get rid of PR x)

I myself am actually going to try the previously proposed [Redirector](https://einaregilsson.com/redirector/) solution as I know regex :sunglasses: and it allows more fine-grained control. (Thanks @egad13! seems like I'll disable PR after all :D)
